compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -0,0 +1,707 @@
+#![allow(unused_imports)]
+use std::str::FromStr;
+use std::string::String;
+
+use rustc_ast::expand::autodiff_attrs::{
+    is_fwd, is_rev, valid_input_activity, valid_ty_for_activity, AutoDiffAttrs, DiffActivity,
+    DiffMode,
+};
+use rustc_ast::ptr::P;
+use rustc_ast::token::{Token, TokenKind};
+use rustc_ast::tokenstream::*;
+use rustc_ast::visit::AssocCtxt::*;
+use rustc_ast::Safety::Unsafe;
+use rustc_ast::{
+    self as ast, AssocItemKind, BindingMode, ByRef, Fn, FnHeader, FnRetTy, FnSig, Generics,
+    ItemKind, MetaItemKind, NestedMetaItem, PatKind, Stmt, StmtKind, TyKind,
+};
+use rustc_expand::base::{Annotatable, ExtCtxt};
+use rustc_span::symbol::{kw, sym, Ident};
+use rustc_span::{Span, Symbol};
+use thin_vec::{thin_vec, ThinVec};
+use tracing::trace;
+
+use crate::errors;
+
+#[cfg(not(not(llvm_enzyme)))]
+pub fn expand(
+    ecx: &mut ExtCtxt<'_>,
+    _expand_span: Span,
+    meta_item: &ast::MetaItem,
+    item: Annotatable,
+) -> Vec<Annotatable> {
+    ecx.sess.dcx().emit_err(errors::AutoDiffSupportNotBuild { span: meta_item.span });
+    return vec![item];
+}
+
+#[cfg(not(llvm_enzyme))]
+fn first_ident(x: &NestedMetaItem) -> rustc_span::symbol::Ident {
+    let segments = &x.meta_item().unwrap().path.segments;
+    assert!(segments.len() == 1);
+    segments[0].ident
+}
+
+#[cfg(not(llvm_enzyme))]
+fn name(x: &NestedMetaItem) -> String {
+    first_ident(x).name.to_string()
+}
+
+#[cfg(not(llvm_enzyme))]
+pub fn from_ast(
+    ecx: &mut ExtCtxt<'_>,
+    meta_item: &ThinVec<NestedMetaItem>,
+    has_ret: bool,
+) -> AutoDiffAttrs {
+    let mode = name(&meta_item[1]);
```

Comment:
Suggestion: I think it would be very helpful if we also explained what's the significance of ordering of items in the attribute, e.g. is it something like

```
[ differential_name, mode, args.., ret_arg ]
```

or something?

Reply:
Remark: in general, I think we want to try avoid explicit array indexing like `meta_item[1]` if possible and favor `split` etc. where possible (and try not to `.unwrap()`) with very explicit panics or diagnostic emissions (as in, are these theorectically unreachable because checked earlier, or unimplemented, or user error etc.).